substrate/frame/revive/src/exec.rs

```diff
@@ -2192,40 +2248,56 @@ where
 
 	fn gas_left(&self) -> u64 {
 		let frame = self.top_frame();
-		if let Some((encoded_len, base_weight)) = self.exec_config.collect_deposit_from_hold {
-			// when using the txhold we know the overall available fee by looking at the tx credit
+
+		// when using the txhold we know the overall available fee by looking at the tx credit
+		// we need to use that to limit gas_left because in that case no storage deposit limit is
+		// set
+		let max_by_credit_hold = if let Some((encoded_len, base_weight)) =
+			self.exec_config.collect_deposit_from_hold
+		{
 			// we work backwards: the gas_left is the overall fee minus what was already consumed
+			let (deposit_consumed, weight_consumed) =
+				self.frames.iter().chain(core::iter::once(&self.first_frame)).fold(
+					(StorageDeposit::default(), Weight::default()),
+					|(deposit, weight), frame| {
+						(
+							deposit.saturating_add(&frame.nested_storage.consumed()),
+							weight.saturating_add(frame.nested_gas.gas_consumed()),
+						)
+					},
+				);
 			let weight_fee_consumed = T::FeeInfo::tx_fee_from_weight(
 				encoded_len,
-				&frame.nested_gas.gas_consumed().saturating_add(base_weight),
+				&weight_consumed.saturating_add(base_weight),
 			);
 			let available = T::FeeInfo::remaining_txfee().saturating_sub(weight_fee_consumed);
-			let deposit_consumed = self
-				.frames
-				.iter()
-				.chain(core::iter::once(&self.first_frame))
-				.fold(StorageDeposit::default(), |acc, frame| {
-					acc.saturating_add(&frame.nested_storage.consumed())
-				});
 			deposit_consumed.available(&available)
 		} else {
-			// when not using the hold we expect the transaction to contain a limit for the storage
-			// deposit we work forwards: add up what is left from both meters
-			// in case no storage limit is set we limit by all the free balance of the signer
+			BalanceOf::<T>::max_value()
+		};
+
+		// this is the actual limit derived from what is set in the meter
+		let max_by_meter = {
 			use frame_support::traits::tokens::{Fortitude, Preservation};
-			let weight_fee_available =
-				T::FeeInfo::weight_to_fee(&frame.nested_gas.gas_left(), Combinator::Min);
-			let available_balance = self
+			let weight_fee_available = T::FeeInfo::weight_to_fee(&frame.nested_gas.gas_left());
+
+			// in the dry run no deposit limit is set.
+			// this means its only limited by the origins free balance
+			let deposit_available = self
 				.origin
 				.account_id()
 				.map(|acc| {
 					T::Currency::reducible_balance(acc, Preservation::Preserve, Fortitude::Polite)
```

Comment:
Why do we use the `reducible_balance` of the origin here? This has no meaning if `exec_config` is ethereum like (i.e., `self.exec_config.collect_deposit_from_hold` is `Some`).

Shouldn't we instead limit the initial storage deposit limit for all substrate like executions by initially checking that this limit is at most the origin's reducible balance – or better even, reduce the origin's balance initially by this deposit limit for substrate like transactions?